src/flint/types/fmpz_mod_mpoly.pyx

```diff
@@ -0,0 +1,1321 @@
+
+        res = []
+        for i in range(len(self)):
+            fmpz_mod_mpoly_get_term_exp_fmpz(vec.double_indirect, self.val, i, self.ctx.val)
+            res.append(vec.to_tuple())
+
+        return res
+
+    def coeffs(self):
+        """
+        Return the coefficients of each term as a fmpz
+
+            >>> from flint import Ordering
+            >>> ctx = fmpz_mod_mpoly_ctx.get_context(2, Ordering.lex, 11, 'x')
+            >>> f = ctx.from_dict({(0, 0): 1, (1, 0): 2, (0, 1): 3, (1, 1): 4})
+            >>> f.coeffs()
+            [4, 2, 3, 1]
+
+        """
+        cdef:
+            fmpz coeff
+            slong i
+
+        res = []
+        for i in range(len(self)):
+            coeff = fmpz.__new__(fmpz)
+            fmpz_mod_mpoly_get_term_coeff_fmpz(coeff.val, self.val, i, self.ctx.val)
+            res.append(coeff)
+
+        return res
+
+    # def terms(self):
+    #     """
+    #     Return the terms of this polynomial as a list of fmpz_mod_mpolys.
+
+    #         >>> from flint import Ordering
+    #         >>> ctx = fmpz_mod_mpoly_ctx.get_context(2, Ordering.lex, 11, 'x')
+    #         >>> f = ctx.from_dict({(0, 0): 1, (1, 0): 2, (0, 1): 3, (1, 1): 4})
+    #         >>> f.terms()
+    #         [4*x0*x1, 2*x0, 3*x1, 1]
+
+    #     """
+    #     cdef:
+    #         fmpz_mod_mpoly term
+    #         slong i
+
+    #     res = []
+    #     for i in range(len(self)):
+    #         term = create_fmpz_mod_mpoly(self.ctx)
+    #         fmpz_mod_mpoly_get_term(term.val, self.val, i, self.ctx.val)
+    #         res.append(term)
+
+    #     return res
+
+    def subs(self, dict_args) -> fmpz_mod_mpoly:
+        """
+        Partial evaluate this polynomial with select constants. Keys must be generator names or generator indices,
+        all values must be fmpz.
+
+            >>> from flint import Ordering
+            >>> ctx = fmpz_mod_mpoly_ctx.get_context(2, Ordering.lex, 11, 'x')
+            >>> f = ctx.from_dict({(0, 0): 1, (1, 0): 2, (0, 1): 3, (1, 1): 4})
+            >>> f.subs({"x1": 0})
+            2*x0 + 1
+
+        """
+        cdef:
+            fmpz_mod_mpoly res
+            slong i
+
+        args = tuple((self.ctx.variable_to_index(k), v) for k, v in dict_args.items())
+
+        # Partial application with args in Z. We evaluate the polynomial one variable at a time
+        res = create_fmpz_mod_mpoly(self.ctx)
+
+        fmpz_mod_mpoly_set(res.val, self.val, self.ctx.val)
+        for i, arg in args:
+            if typecheck(arg, fmpz_mod):
+                fmpz_mod_mpoly_evaluate_one_fmpz(res.val, res.val, i, (<fmpz_mod>arg).val, self.ctx.val)
+            else:
+                arg_fmpz = any_as_fmpz(arg)
+                if arg_fmpz is NotImplemented:
+                    raise TypeError(f"cannot coerce {type(arg)} to fmpz")
+                fmpz_mod_mpoly_evaluate_one_fmpz(res.val, res.val, i, (<fmpz>arg).val, self.ctx.val)
```

Comment:
I think it should be `(<fmpz>arg_fmpz).val` here.

Otherwise this is casting a PyLong to fmpz I guess which doesn't work.

The tests pass under Python 3.12 if this is changed to `arg_fmpz`.

Reply:
It certainly should be, thanks.